src/pod.rs

```diff
@@ -60,6 +56,216 @@ pub(crate) fn bytes_of<T: Pod>(val: &T) -> &[u8] {
+        self.skip(offset)?;
+        self.read_bytes(count)
+    }
+
+    #[inline]
+    pub fn read<T: Pod>(&mut self) -> Option<&'data T> {
+        match from_bytes(self.0) {
+            Some((value, tail)) => {
+                self.0 = tail;
+                Some(value)
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_at<T: Pod>(mut self, offset: usize) -> Option<&'data T> {
+        self.skip(offset)?;
+        self.read()
+    }
+
+    #[inline]
+    pub fn read_slice<T: Pod>(&mut self, count: usize) -> Option<&'data [T]> {
+        match slice_from_bytes(self.0, count) {
+            Some((value, tail)) => {
+                self.0 = tail;
+                Some(value)
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_slice_at<T: Pod>(mut self, offset: usize, count: usize) -> Option<&'data [T]> {
+        self.skip(offset)?;
+        self.read_slice(count)
+    }
+
+    /// Read a null terminated string.
+    ///
+    /// Does not assume any encoding.
+    /// Reads past the null byte, but doesn't return it.
+    #[inline]
+    pub fn read_string(&mut self) -> Option<&'data [u8]> {
+        match self.0.iter().position(|&x| x == 0) {
+            Some(null) => {
+                // These will never fail.
+                let bytes = self.read_bytes(null)?;
+                self.skip(1)?;
+                Some(bytes.0)
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    /// Read a null terminated string at an offset.
+    ///
+    /// Does not assume any encoding. Does not return the null byte.
+    #[inline]
+    pub fn read_string_at(mut self, offset: usize) -> Option<&'data [u8]> {
+        self.skip(offset)?;
+        self.read_string()
+    }
+}
+
+/// A newtype for byte vectors.
+///
+/// It provides convenience methods for `Pod` types.
+// TODO: should this be an extension trait for `Vec<u8>` instead?
+#[derive(Default, Clone, PartialEq, Eq)]
+pub(crate) struct BytesMut(pub Vec<u8>);
+
+impl fmt::Debug for BytesMut {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        debug_list_bytes(&self.0, fmt)
+    }
+}
+
+impl BytesMut {
+    #[inline]
+    pub fn new() -> Self {
+        BytesMut(Vec::new())
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    #[inline]
+    pub fn resize(&mut self, new_len: usize, value: u8) {
+        self.0.resize(new_len, value);
+    }
+
+    #[inline]
+    pub fn write<T: Pod>(&mut self, val: &T) {
+        self.0.extend_from_slice(bytes_of(val))
+    }
+
+    #[inline]
+    pub fn write_at<T: Pod>(&mut self, offset: usize, val: &T) -> Option<()> {
```

Comment:
we'll want callers to deal with `None` failures returned by this function, rather than assuming they succeeded, so it probably makes sense to use `Result` here (unlike the other methods, which return a read value or don't, this isn't returning a value that's used directly, so to speak, so I think the risk is higher).

Might as well have `Result<(), BytesError>` or something where `BytesError` is just `struct BytesError { _: () }` or something, so that we get the usage checking from `rustc`.

Reply:
Good point. I changed it to `Result<(), ()>`, I think that's ok for now?

Probably want to change the read methods too at some point, but I'm still not sure what the read error handling should look like for this crate (or is `&static str` good enough?).

Reply:
Sounds good -- all these methods are currently internal only so using `()` as the error type seems fine. Eventually, if we start exposing these details or bubbling them up, we might want a custom error type, but until then not worth fretting over.